### PR TITLE
[processor/k8sattribute] support attribute k8s.deployment.uid

### LINKED ISSUE #12936: [processor/k8sattributes] k8s.deployment.name attribute can be set to a wrong value
`k8s.deployment.name` attribute value is taken from the pod name using regexp. This is not a robust approach. In some scenarios it can set the attribute to a wrong value. For example a Stateful Set called `sts-test` will create a pod called `sts-test-1`, telemetry from that pod will get the attribute `k8s.deployment.name: sts` which is incorrect, `k8s.deployment.name` should not be set for such pod.

There are two ways to solve the problem:

1. Make the [regexp](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/k8sattributesprocessor/internal/kube/client.go#L64) more strict. This will lower chances for getting a wrong attribute, but it's impossible to completely eliminate the chances with the regexp approach

2. Use the k8s API to fetch an owner of the pod, check if it's replicaset controlled by a deployment, then set the correct value for `k8s.deplyment.name` attribute. This is a robust approach, but we need to make sure it doesn't introduce a significant computational overhead and load on k8s API.

### PR DESCRIPTION
**Pull request by fatsheep9146**
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

fix #12936

### CODE REVIEW

**Comment by fatsheep9146** [comment]
@dmitryax The pr is ready for review.

**Comment by fatsheep9146** [comment]
ping @dmitryax, could you help review this?

**Review by povilasv — COMMENTED** [review]
I think maybe instead of going this route, we should discover ReplicaSets and get rid of Deployment regex?

The idea behind this, is that regex can match things that are not managed by Deployment, and we would still set Deployment name incorrectly.

So I think we should instead discover ReplicaSets and parse the Pod's Owner References to find the ReplicaSet. Example Pod Owner references:

```
apiVersion: v1
kind: Pod
metadata:
  labels:
    app.kubernetes.io/instance: opentelemetry-collector
    app.kubernetes.io/name: opentelemetry-collector
    component: standalone-collector
    pod-template-hash: 6c45f8d6f6
  name: opentelemetry-collector-6c45f8d6f6-s4jq4
  ownerReferences:                                                  
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true 
    kind: ReplicaSet
    name: opentelemetry-collector-6c45f8d6f6
    uid: ee75293d-14ec-42a0-9548-a768d9e07c48
 ```
 
 And from ReplicaSet's Owner Reference we can get Deployment name, and id:
 
 ```
 apiVersion: apps/v1
kind: ReplicaSet
metadata:
  labels:
    app.kubernetes.io/instance: opentelemetry-collector
    app.kubernetes.io/name: opentelemetry-collector
    component: standalone-collector
    pod-template-hash: 6c45f8d6f6
  name: opentelemetry-collector-6c45f8d6f6
  namespace: sys-mon
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: Deployment
    name: opentelemetry-collector
    uid: f875e269-0e76-4735-a02d-a254ed056111
```

**Comment by fatsheep9146** [comment]
> I think maybe instead of going this route, we should discover ReplicaSets and get rid of Deployment regex?
> 
> The idea behind this, is that regex can match things that are not managed by Deployment, and we would still set Deployment name incorrectly.
> 
> So I think we should instead discover ReplicaSets and parse the Pod's Owner References to find the ReplicaSet. Example Pod Owner references:
> 
> ```
> apiVersion: v1
> kind: Pod
> metadata:
>   labels:
>     app.kubernetes.io/instance: opentelemetry-collector
>     app.kubernetes.io/name: opentelemetry-collector
>     component: standalone-collector
>     pod-template-hash: 6c45f8d6f6
>   name: opentelemetry-collector-6c45f8d6f6-s4jq4
>   ownerReferences:                                                  
>   - apiVersion: apps/v1
>     blockOwnerDeletion: true
>     controller: true 
>     kind: ReplicaSet
>     name: opentelemetry-collector-6c45f8d6f6
>     uid: ee75293d-14ec-42a0-9548-a768d9e07c48
> ```
> 
> And from ReplicaSet's Owner Reference we can get Deployment name, and id:
> 
> ```
> apiVersion: apps/v1
> kind: ReplicaSet
> metadata:
>  labels:
>    app.kubernetes.io/instance: opentelemetry-collector
>    app.kubernetes.io/name: opentelemetry-collector
>    component: standalone-collector
>    pod-template-hash: 6c45f8d6f6
>  name: opentelemetry-collector-6c45f8d6f6
>  namespace: sys-mon
>  ownerReferences:
>  - apiVersion: apps/v1
>    blockOwnerDeletion: true
>    controller: true
>    kind: Deployment
>    name: opentelemetry-collector
>    uid: f875e269-0e76-4735-a02d-a254ed056111
> ```

I think this make senses, and it is the most accurate way to implement this, I'll change like this.
@povilasv @dmitryax

**Comment by fatsheep9146** [comment]
I will start fixing this

**Comment by fatsheep9146** [comment]
@povilasv I already changed to replicaset, could you help review about this again?

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 1]
```diff
@@ -639,3 +657,63 @@ func (c *WatchClient) extractNamespaceLabelsAnnotations() bool {
...
+	}
+
+	c.m.Lock()
+	if replicaset.Name != "" {
+		c.Replicasets[string(replicaset.UID)] = newReplicaset
+	}
+	c.m.Unlock()
+}
+
+// GetReplicaset takes replicaset uild
```
Comment (evan-bradley) [inline processor/k8sattributesprocessor/internal/kube/client.go:710]: I suspect `uild` is a typo of `uid`, but if not, could you clear up this comment? If it is a typo, I think the comment can probably be removed.

File: processor/k8sattributesprocessor/internal/kube/kube.go [hunk 2]
```diff
@@ -297,3 +297,17 @@ type AssociationSource struct {
...
 }
+
+// Deployment represents a kubernetes deployment.
+type Deployment struct {
+	Name string
+	UID  string
+}
+
+// Replicaset represents a kubernetes replicaset.
+type Replicaset struct {
```
Comment (evan-bradley) [inline processor/k8sattributesprocessor/internal/kube/kube.go:308]: Nitpick: this should be `ReplicaSet`, which is how Kubernetes writes it.

**Review by evan-bradley — COMMENTED** [review]
Looks mostly good to me, just a few comments. Nice work with tracking the relationship between Pods and Deployments using ReplicaSets, I think that's definitely an improvement.

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 1 above)
Comment (fatsheep9146) [reply processor/k8sattributesprocessor/internal/kube/client.go]: It's typo. =D

File: processor/k8sattributesprocessor/internal/kube/client_test.go [hunk 3]
```diff
@@ -203,6 +235,11 @@ func TestNamespaceAdd(t *testing.T) {
 	namespaceAddAndUpdateTest(t, c, c.handleNamespaceAdd)
 }
 
+func TestReplicasetAdd(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (fatsheep9146) [reply processor/k8sattributesprocessor/internal/kube/client_test.go]: I combine adding, updating, deleting tests in one test `TestReplicaSetHandler`

**Comment by fatsheep9146** [comment]
could you help review this again? @evan-bradley

**Review by evan-bradley — COMMENTED** [review]
Looks mostly good to me, just a few nitpicks mainly around spelling.

**Comment by fatsheep9146** [comment]
could you help review this again? @evan-bradley

**Comment by fatsheep9146** [comment]
could you help review this again? @evan-bradley @povilasv @djaglowski

**Review by povilasv — APPROVED** [review]
Other than that LGTM, but I am not a maintainer :)

**Comment by fatsheep9146** [comment]
> Other than that LGTM, but I am not a maintainer :)

Any reviews are all welcome, thank you.  =D

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 4]
```diff
@@ -639,3 +657,71 @@ func (c *WatchClient) extractNamespaceLabelsAnnotations() bool {
...
+	} else {
+		c.logger.Error("object received was not of type apps_v1.ReplicaSet", zap.Any("received", obj))
+	}
+}
+
+func (c *WatchClient) addOrUpdateReplicaSet(replicaset *apps_v1.ReplicaSet) {
+	var deployment *Deployment
+	for _, ownerReference := range replicaset.OwnerReferences {
+		if ownerReference.Kind == "Deployment" {
+			deployment = &Deployment{
```
Comment (evan-bradley) [inline processor/k8sattributesprocessor/internal/kube/client.go:695]: This is hopefully a rare case, but can we add a warn log here if we find that another deployment owner reference has been found? I think that may cause strange and confusing data if it were to happen.

Alternatively, if we know for sure that Kubernetes guarantees a ReplicaSet will only ever have a single Deployment as an owner, then we should break out of the loop and have a comment documenting that there can only be one reference.

Comment (fatsheep9146) [reply processor/k8sattributesprocessor/internal/kube/client.go:695]: After some research, I think we can use `ownerReference.controller` which denotes whether this controller take over the control of replicaset. The design doc is right [here](https://github.com/kubernetes/design-proposals-archive/blob/main/api-machinery/controller-ref.md)

I also made a test, I can create a replicaset which ownerReferences have two deployments in it, but only one `ownerReference.controller` is true.
```
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  generation: 1
  labels:
    app: test2
  name: test2-7c9d5954f6
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: Deployment
    name: test2
    uid: 2ab30db9-c151-4db4-bc43-680bfcfba696
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: false
    kind: Deployment
    name: test1
    uid: 13ca1c8b-5c96-4ec0-a3af-0e08110e22dc
```

If I set both controller to true, the apiserver will deny the request with error message.
```
Only one reference can have Controller set to true. Found "true" in references for test2 and test1
```

**Comment by fatsheep9146** [comment]
Could you help review this again? @evan-bradley

**Review by evan-bradley — APPROVED** [review]
Looks good to me. Great find with the controller flag in the owner references.

**Comment by fatsheep9146** [comment]
do you have time to review this pr ? @dmitryax

**Comment by dmitryax** [comment]
Sorry for the delay. Given that we got some significant changes to this processor that introduces bugs, I'm a bit skeptical about rushing to merge this one. 

We discussed with @povilasv plans about introducing e2e tests for k8s related functionality of the collector. Can we delay this PR until the tests are introduced? Let me know if you want to help with that.

**Comment by dmitryax** [comment]
@fatsheep9146 I filed an issue to start with

**Comment by swiatekm-sumo** [comment]
@dmitryax @fatsheep9146 looks like we now have E2E tests for the k8sattributes processor, so can we revisit this PR?

Having skimmed the code, I have two concerns about the current change:
* the Replicaset informer is enabled unconditionally, so users who don't need it will pay the cost regardless
* the `WatchClient` is growing a bit large - we might want to consider moving the responsibility of managing the owner data into a separate component

I don't think these are obstacles to merging this though.

**Comment by dmitryax** [comment]
> the Replicaset informer is enabled unconditionally, so users who don't need it will pay the cost regardless

I'd say at least this one should be addressed in this PR. We should avoid putting extra pressure on k8s API and compute resources for the functionality that isn't being used

**Comment by fatsheep9146** [comment]
> > the Replicaset informer is enabled unconditionally, so users who don't need it will pay the cost regardless
> 
> I'd say at least this one should be addressed in this PR. We should avoid putting extra pressure on k8s API and compute resources for the functionality that isn't being used

I already solved this one.

About another improvement, I think I can submit another pr, since this one already has many modifications.

**Review by swiatekm — APPROVED** [review]
LGTM

**Comment by fatsheep9146** [comment]
@swiatekm-sumo Thx! The failed test are still needed to be resolved, I'm already doing this.

**Comment by fatsheep9146** [comment]
@dmitryax The pr is ready to be reviewed.
The failed test is not related to this pr.

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 5]
```diff
@@ -661,3 +696,71 @@ func needContainerAttributes(rules ExtractionRules) bool {
...
+	newReplicaSet := &ReplicaSet{
+		Name:      replicaset.Name,
+		Namespace: replicaset.Namespace,
+		UID:       string(replicaset.UID),
+	}
+
+	var deployment *Deployment
+	for _, ownerReference := range replicaset.OwnerReferences {
+		if ownerReference.Kind == "Deployment" && ownerReference.Controller != nil && *ownerReference.Controller {
+			deployment = &Deployment{
```
[the remark this answers is not in the document]
Comment (fatsheep9146) [reply processor/k8sattributesprocessor/internal/kube/client.go:740]: That make senses

File: processor/k8sattributesprocessor/internal/kube/kube.go [hunk 6]
```diff
@@ -309,3 +310,17 @@ type AssociationSource struct {
...
+	Name string
+	UID  string
+}
+
+// ReplicaSet represents a kubernetes replicaset.
+type ReplicaSet struct {
+	Name       string
+	Namespace  string
+	UID        string
+	Deployment *Deployment
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/kube.go:325]: Why pointer? It should be good to use the value and avoid the allocation

**Review by dmitryax — APPROVED** [review]
Just one nit, otherwise LGTM

**Review by dmitryax — APPROVED** [review]

**Pull request merged**